pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -40,17 +41,62 @@ type VolumeRestrictions struct {
 var _ framework.PreFilterPlugin = &VolumeRestrictions{}
 var _ framework.FilterPlugin = &VolumeRestrictions{}
 var _ framework.EnqueueExtensions = &VolumeRestrictions{}
-
-// Name is the name of the plugin used in the plugin registry and configurations.
-const Name = names.VolumeRestrictions
+var _ framework.StateData = &preFilterState{}
 
 const (
+	// Name is the name of the plugin used in the plugin registry and configurations.
+	Name = names.VolumeRestrictions
+	// preFilterStateKey is the key in CycleState to VolumeRestrictions pre-computed data for Filtering.
+	// Using the name of the plugin will likely help us avoid collisions with other plugins.
+	preFilterStateKey = "PreFilter" + Name
+
 	// ErrReasonDiskConflict is used for NoDiskConflict predicate error.
 	ErrReasonDiskConflict = "node(s) had no available disk"
 	// ErrReasonReadWriteOncePodConflict is used when a pod is found using the same PVC with the ReadWriteOncePod access mode.
 	ErrReasonReadWriteOncePodConflict = "node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode"
 )
 
+// preFilterState computed at PreFilter and used at Filter.
+type preFilterState struct {
+	// Names of the pod's volumes using the ReadWriteOncePod access mode.
+	readWriteOncePodPVCs sets.Set[string]
+	// The number of pods that are also using these ReadWriteOncePod volumes.
+	numConflictingPods int
+}
+
+func (s *preFilterState) updateWithPod(podInfo *framework.PodInfo, multiplier int64) {
+	if s.hasConflict(podInfo) {
+		if multiplier > 0 {
+			s.numConflictingPods += 1
+		} else {
+			s.numConflictingPods -= 1
+		}
+	}
+}
+
+func (s *preFilterState) hasConflict(podInfo *framework.PodInfo) bool {
```

Comment:
shouldn't this return how many volumes are conflicting?

add a unit or integration test that exercises preemption when there are more than one conflict

Reply:
I'm not convinced that this works, because in `calPreFilterState` you sum up the number of pods that conflict for each PV.

And in updateWithPod, you only count the pods if there is a conflict, independently from the number of pvcs that this pod is having a conflict with.

Reply:
`calPreFilterState` had a bug where pods were double-counted when they share more than one conflicting PVC. I think if we now have an accurate count of `numConflictingPods` in PreFilter, then we should be able to increment and decrement it by 1 (pod) and not worry about the number of conflicting PVCs

Reply:
See the most recent commit for context. Included is an integration test that caught this double-counting issue.

Reply:
```
=== RUN   TestReadWriteOncePodPreemption/preempt_single_pod
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 0
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/preemptor-pod
numConflictingPods is 1
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 0
adding pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 1
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 0

=== RUN   TestReadWriteOncePodPreemption/preempt_two_pods
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-1
numConflictingPods is 0
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-2
numConflictingPods is 0
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/preemptor-pod
numConflictingPods is 2
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-1
numConflictingPods is 1
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-2
numConflictingPods is 0
adding pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-2
numConflictingPods is 1
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-2
numConflictingPods is 0
adding pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-1
numConflictingPods is 1
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-1
numConflictingPods is 0
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/preemptor-pod
numConflictingPods is 2

^  not sure why this happens yet but I think it's because the preemptor-pod is re-scheduled so it re-calculates the preFilterState

=== RUN   TestReadWriteOncePodPreemption/preempt_single_pod_with_two_volumes
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 0
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/preemptor-pod
numConflictingPods is 1
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 0
adding pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 1
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 0
```